### PR TITLE
Improve error reporting for #'protocol-method-var?

### PR DESCRIPTION
**Pull request by thumbnail**
## Brief
I noticed users struggled decypering with the "this is not a protocol-method"-error; especially when paired with speced.def/defprotocol. 

By using a set, we get a more descriptive expound explanation. 

### from
image

### to
image

## QA plan

n/a

## Author checklist

* [x] I have QAed the functionality
* [x] The PR has a reasonably reviewable size and a meaningful commit history
* [x] I have run the [branch formatter](https://github.com/nedap/formatting-stack/blob/332a419034ab46fad526a5592f4257353bd695b6/src/formatting_stack/branch_formatter.clj) and observed no new/significative warnings
* [x] The build passes
* [x] I have self-reviewed the PR prior to assignment
* Additionally, I have code-reviewed iteratively the PR considering the following aspects in isolation:
  * [x] Correctness
  * [x] Robustness (red paths, failure handling etc)
  * [x] Test coverage
  * [x] Spec coverage
  * [x] Documentation
  * [ ] Security
  * [ ] Performance
  * [x] Breaking API changes
  * [x] Cross-compatibility (Clojure/ClojureScript)

## Reviewer checklist

* [ ] I have checked out this branch and reviewed it locally, running it
* [ ] I have QAed the functionality
* [x] I have reviewed the PR
* Additionally, I have code-reviewed iteratively the PR considering the following aspects in isolation:
  * [ ] Correctness
  * [ ] Robustness (red paths, failure handling etc)
  * [ ] Test coverage
  * [ ] Spec coverage
  * [ ] Documentation
  * [ ] Security
  * [ ] Performance
  * [ ] Breaking API changes
  * [ ] Cross-compatibility (Clojure/ClojureScript)

### CODE REVIEW

File: src/nedap/utils/modular/impl/implement.cljc [hunk 1]
```diff
@@ -42,16 +42,13 @@
...
-  [v]
-  (and (-> v meta :protocol)
-       (-> v deref fn?)
-       (let [{:keys [on] :as protocol} (-> v meta :protocol deref)]
-         (if (-> protocol :extend-via-metadata)
-           true
-           (throw (ex-info "The targeted protocol does not have `:extend-via-metadata` activated."
-                           {:protocol on}))))))
+(defn ns-protocol-method-vars
+  "Returns a set of every abstract protocol in `ns`"
```
Comment (vemv) [inline src/nedap/utils/modular/impl/implement.cljc:46]: * "every abstract protocol **method**"
* Be wary of losing the `fn?` check - for one thing it would be an unrelated change
* Try prefering `val` over `second`, it's clearer
* Maybe replace the docstring with a spec
* Seems a transduceable transformation

**Review by vemv — COMMENTED** [review]
Thanks for this one!

File: src/nedap/utils/modular/impl/implement.cljc [hunk 2]
```diff
@@ -42,16 +42,13 @@
...
+  [ns]
+  (->> (ns-publics ns)
```
Comment (vemv) [inline src/nedap/utils/modular/impl/implement.cljc:48]: `(defprotocol ^:private foo (bar [x]))` will escape `ns-publics`

Comment (thumbnail) [reply src/nedap/utils/modular/impl/implement.cljc:48]: I'm not sure what the behaviour is for implementing private protocols (either from the same ns, or from another ns, or directly as a var etc.);

I assume that you can implement private protocols in the same ns, not outside. If that's true `ns-interns` needs to be used for the ns in which `implements` is called (`ns-publics` otherwise).

I'd need to experiment a bit.

**Review by tcoenraad — APPROVED** [review]

File: src/nedap/utils/modular/impl/implement.cljc [hunk 3]
```diff
@@ -42,16 +43,14 @@
...
-         (if (-> protocol :extend-via-metadata)
-           true
-           (throw (ex-info "The targeted protocol does not have `:extend-via-metadata` activated."
-                           {:protocol on}))))))
+(speced/defn ^{::speced/spec (spec/coll-of var? :kind set?)} ns-protocol-method-vars
+  [ns]
+  (into #{}
+        (comp (map val)
+              (filter (comp :protocol meta))
+              (filter (comp #{ns} #(.ns %)))
```
Comment (vemv) [inline src/nedap/utils/modular/impl/implement.cljc:51]: are you sure that dot access is needed? there are many defns suffixed `ns` or `namespace` in clojure.core

File: test/unit/nedap/utils/modular/api/implement.cljc [hunk 4]
```diff
@@ -68,11 +69,13 @@
      (testing "made-up symbols"
 
        ;; no `are` here since `eval '` makes it harder to predict what will actually happen
-       (is (spec-assertion-thrown? `sut.impl/protocol-method-var? (eval '(nedap.utils.modular.api/implement {}
-                                                                           foo-impl oooooommmmmggggg))))
+       (is (spec-assertion-thrown? #{#'bar #'foo}
+                                   (eval `(nedap.utils.modular.api/implement {}
+                                            foo-impl oooooommmmmggggg))))
 
-       (is (spec-assertion-thrown? `sut.impl/protocol-method-var? (eval '(nedap.utils.modular.api/implement {}
```
Comment (vemv) [inline test/unit/nedap/utils/modular/api/implement.cljc:74]: why was single-quote changed to backtick? It's not the same, `oooooommmmmggggg` will be now ns-qualifed prior to evaluation

File: src/nedap/utils/modular/impl/implement.cljc (hunk 3 above)
Comment (thumbnail) [reply src/nedap/utils/modular/impl/implement.cljc:51]: yes; `%` refers to a Var, i am not aware of a different way of getting the ns of those.

File: test/unit/nedap/utils/modular/api/implement.cljc (hunk 4 above)
Comment (thumbnail) [reply test/unit/nedap/utils/modular/api/implement.cljc:74]: It caused compile errors on the first load of the code. After a refresh it was fine.

To avoid a goosechase I asserted the implementation was still fine, and decided to change the quote. If there's any issue with this change feel free to open an issue.

File: src/nedap/utils/modular/impl/implement.cljc [hunk 5]
```diff
@@ -42,16 +43,14 @@
...
-  [v]
-  (and (-> v meta :protocol)
-       (-> v deref fn?)
-       (let [{:keys [on] :as protocol} (-> v meta :protocol deref)]
-         (if (-> protocol :extend-via-metadata)
-           true
-           (throw (ex-info "The targeted protocol does not have `:extend-via-metadata` activated."
-                           {:protocol on}))))))
+(speced/defn ^{::speced/spec (spec/coll-of var? :kind set?)} ns-protocol-method-vars
+  [ns]
```
[the remark this answers is not in the document]
Comment (thumbnail) [reply src/nedap/utils/modular/impl/implement.cljc:47]: given that `ns` and `namespace` are both used, and every `ns-*` fn in clojure.core also shadows `ns` i decided this is fine. Esp. because the `ns`-macro will never be inside another form anyway.
